File: dsa-problems/Graphs/solutions/python/GRP-002-lab-network-dfs.py
import sys
sys.setrecursionlimit(200000)
from typing import List

def dfs_traversal(n: int, adj: List[List[int]]) -> List[int]:
    """
    Perform DFS traversal starting from node 0.
    
    Args:
        n: Number of nodes
        adj: Adjacency list representation
    
    Returns:
        List of nodes in preorder DFS visitation order
    """
    result = []

    # Sort neighbors for deterministic traversal
    for neighbors in adj:
        neighbors.sort()

    visited = [False] * n
    
    def dfs(node):
        # Mark as visited and add to result (preorder)
        visited[node] = True
        result.append(node)
        
        # Recursively visit all unvisited neighbors
        for neighbor in adj[node]:
            if not visited[neighbor]:
                dfs(neighbor)
    
    # Start DFS from node 0
    dfs(0)
    
    return result

def main():
    n = int(input())
    m = int(input())
    
    adj = [[] for _ in range(n)]
    
    for _ in range(m):
        u, v = map(int, input().split())
        adj[u].append(v)
        adj[v].append(u)
    
    result = dfs_traversal(n, adj)
    print(' '.join(map(str, result)))

if __name__ == "__main__":
    main()

File: dsa-problems/Graphs/solutions/cpp/GRP-002-lab-network-dfs.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
private:
    vector<int> result;
    vector<bool> visited;
    
    void dfs(int node, vector<vector<int>>& adj) {
        // Mark as visited and add to result (preorder)
        visited[node] = true;
        result.push_back(node);
        
        // Recursively visit all unvisited neighbors
        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor, adj);
            }
        }
    }
    
public:
    vector<int> dfsTraversal(int n, vector<vector<int>>& adj) {
        result.clear();
        visited.assign(n, false);
        
        // Start DFS from node 0
        dfs(0, adj);
        
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n, m;
    cin >> n >> m;
    
    vector<vector<int>> adj(n);
    
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    // Sort neighbors for deterministic traversal
    for (int i = 0; i < n; i++) {
        sort(adj[i].begin(), adj[i].end());
    }

    Solution solution;
    vector<int> result = solution.dfsTraversal(n, adj);
    
    for (int i = 0; i < result.size(); i++) {
        cout << result[i];
        if (i < result.size() - 1) cout << " ";
    }
    cout << "\n";
    
    return 0;
}
